From 31fc7cc82ce8e030fc7864fb00879c20bf5c40d5 Mon Sep 17 00:00:00 2001
From: Zheng Zucheng <zhengzucheng@huawei.com>
Date: Sat, 20 May 2023 20:44:45 +0800
Subject: [PATCH 235/780] [Huawei] sched: Add normalize_usage interface to show
 normalize usage

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188931

--------------------------------

Add /sys/fs/cgroup/cpu/xx/cpuacct.normalize_usage to show normalize usage
based on the actual computing power, unit is nanosecond;
user can 'echo 0 > /sys/fs/cgroup/cpu/xx/cpuacct.normalize_usage' to clean
historical data.

Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 include/linux/cgroup.h |   7 ++
 kernel/sched/cpuacct.c | 157 +++++++++++++++++++++++++++++++++++++++++
 2 files changed, 164 insertions(+)

diff --git a/include/linux/cgroup.h b/include/linux/cgroup.h
index 5b60dbad837c..841b4ef240f0 100644
--- a/include/linux/cgroup.h
+++ b/include/linux/cgroup.h
@@ -772,6 +772,10 @@ static inline void cpuacct_account_field(struct task_struct *tsk, int index,
 					 u64 val) {}
 #endif
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+void cpuacct_normalize_charge(struct task_struct *tsk, u64 cputime);
+#endif
+
 void __cgroup_account_cputime(struct cgroup *cgrp, u64 delta_exec);
 void __cgroup_account_cputime_field(struct cgroup *cgrp,
 				    enum cpu_usage_stat index, u64 delta_exec);
@@ -783,6 +787,9 @@ static inline void cgroup_account_cputime(struct task_struct *task,
 
 	cpuacct_charge(task, delta_exec);
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+	cpuacct_normalize_charge(task, delta_exec);
+#endif
 	rcu_read_lock();
 	cgrp = task_dfl_cgroup(task);
 	if (cgroup_parent(cgrp))
diff --git a/kernel/sched/cpuacct.c b/kernel/sched/cpuacct.c
index f48c2dcd8340..a62ccff5ab14 100644
--- a/kernel/sched/cpuacct.c
+++ b/kernel/sched/cpuacct.c
@@ -25,12 +25,24 @@ struct cpuacct_usage {
 	u64	usages[CPUACCT_STAT_NSTATS];
 };
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+struct cpuacct_norm_usage {
+	u64	usages;
+};
+
+static DEFINE_PER_CPU(struct cpuacct_norm_usage, root_cpuacct_norm_cpuusage);
+#endif
+
 /* track CPU usage of a group of tasks and its child groups */
 struct cpuacct {
 	struct cgroup_subsys_state	css;
 	/* cpuusage holds pointer to a u64-type object on every CPU */
 	struct cpuacct_usage __percpu	*cpuusage;
 	struct kernel_cpustat __percpu	*cpustat;
+
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+	struct cpuacct_norm_usage __percpu	*norm_cpuusage;
+#endif
 };
 
 static inline struct cpuacct *css_ca(struct cgroup_subsys_state *css)
@@ -53,6 +65,10 @@ static DEFINE_PER_CPU(struct cpuacct_usage, root_cpuacct_cpuusage);
 static struct cpuacct root_cpuacct = {
 	.cpustat	= &kernel_cpustat,
 	.cpuusage	= &root_cpuacct_cpuusage,
+
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+	.norm_cpuusage	= &root_cpuacct_norm_cpuusage,
+#endif
 };
 
 /* Create a new CPU accounting group */
@@ -76,8 +92,17 @@ cpuacct_css_alloc(struct cgroup_subsys_state *parent_css)
 	if (!ca->cpustat)
 		goto out_free_cpuusage;
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+	ca->norm_cpuusage = alloc_percpu(struct cpuacct_norm_usage);
+	if (!ca->norm_cpuusage)
+		goto out_free_cpustat;
+#endif
 	return &ca->css;
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+out_free_cpustat:
+	free_percpu(ca->cpustat);
+#endif
 out_free_cpuusage:
 	free_percpu(ca->cpuusage);
 out_free_ca:
@@ -91,6 +116,9 @@ static void cpuacct_css_free(struct cgroup_subsys_state *css)
 {
 	struct cpuacct *ca = css_ca(css);
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+	free_percpu(ca->norm_cpuusage);
+#endif
 	free_percpu(ca->cpustat);
 	free_percpu(ca->cpuusage);
 	kfree(ca);
@@ -297,6 +325,127 @@ static int cpuacct_stats_show(struct seq_file *sf, void *v)
 #ifdef CONFIG_QOS_NORMALIZE_UTIL
 unsigned int __read_mostly sysctl_multi_task_factor = 60;
 unsigned int __read_mostly sysctl_single_task_factor = 100;
+
+static u64 cpuusage_sibling_runtime(int cpu)
+{
+	u64 sum;
+	u64 lost = 0;
+	struct rq *rq = NULL;
+	struct smt_peer_stat *pstat = NULL;
+	int sibling = cpu_sibling_cpuid(cpu);
+
+	if (unlikely(!sched_smt_active() || sibling == -1 ||
+		     cpu_is_offline(sibling)))
+		return 0;
+
+	pstat = &per_cpu(smt_util_stat, sibling);
+	rq = cpu_rq(sibling);
+	if (unlikely(!raw_spin_trylock(&rq->lock))) {
+		raw_spin_lock(&pstat->lock);
+		sum = pstat->sum;
+		pstat->sum = 0;
+		raw_spin_unlock(&pstat->lock);
+		return sum;
+	}
+
+	raw_spin_lock(&pstat->lock);
+	if (pstat->last_update_time) {
+		update_rq_clock(rq);
+		lost = rq_clock_task(rq) - pstat->last_update_time;
+		if (unlikely((s64)lost < 0))
+			lost = 0;
+
+		pstat->last_update_time = rq_clock_task(rq);
+	}
+	sum = pstat->sum + lost;
+	pstat->sum = 0;
+	raw_spin_unlock(&pstat->lock);
+	raw_spin_unlock(&rq->lock);
+
+	return sum;
+}
+
+void cpuacct_normalize_charge(struct task_struct *tsk, u64 cputime)
+{
+	u64 delta;
+	u64 ptime;
+	struct cpuacct *ca;
+
+	ptime = cpuusage_sibling_runtime(task_cpu(tsk));
+	if (unlikely(ptime > cputime))
+		ptime = cputime;
+
+	delta = ((cputime - ptime) * sysctl_single_task_factor +
+		 ptime * sysctl_multi_task_factor) / 100;
+	rcu_read_lock();
+	for (ca = task_ca(tsk); ca; ca = parent_ca(ca))
+		__this_cpu_add(ca->norm_cpuusage->usages, delta);
+	rcu_read_unlock();
+}
+
+static u64 cpuacct_normalize_cpuusage_read(struct cpuacct *ca, int cpu)
+{
+	struct cpuacct_norm_usage *norm_cpuusage = NULL;
+	u64 data;
+
+#ifndef CONFIG_64BIT
+	/*
+	 * Take rq->lock to make 64-bit read safe on 32-bit platforms.
+	 */
+	raw_spin_lock_irq(&cpu_rq(cpu)->lock);
+#endif
+
+	norm_cpuusage = per_cpu_ptr(ca->norm_cpuusage, cpu);
+	data = norm_cpuusage->usages;
+
+#ifndef CONFIG_64BIT
+	raw_spin_unlock_irq(&cpu_rq(cpu)->lock);
+#endif
+
+	return data;
+}
+
+static u64 cpuusage_normalize_read(struct cgroup_subsys_state *css,
+				   struct cftype *cft)
+{
+	struct cpuacct *ca = css_ca(css);
+	u64 totalcpuusage = 0;
+	int i;
+
+	for_each_possible_cpu(i)
+		totalcpuusage += cpuacct_normalize_cpuusage_read(ca, i);
+
+	return totalcpuusage;
+}
+
+static int cpuusage_normalize_write(struct cgroup_subsys_state *css,
+				    struct cftype *cft, u64 val)
+{
+	int cpu;
+	struct cpuacct *ca = css_ca(css);
+	struct cpuacct_norm_usage *norm_cpuusage = NULL;
+
+	/*
+	 * Only allow '0' here to do a reset.
+	 */
+	if (val)
+		return -EINVAL;
+
+	for_each_possible_cpu(cpu) {
+		norm_cpuusage = per_cpu_ptr(ca->norm_cpuusage, cpu);
+#ifndef CONFIG_64BIT
+		/*
+		 * Take rq->lock to make 64-bit write safe on 32-bit platforms.
+		 */
+		raw_spin_lock_irq(&cpu_rq(cpu)->lock);
+#endif
+		norm_cpuusage->usages = val;
+#ifndef CONFIG_64BIT
+		raw_spin_unlock_irq(&cpu_rq(cpu)->lock);
+#endif
+	}
+	return 0;
+}
 #endif
 
 static struct cftype files[] = {
@@ -333,6 +482,14 @@ static struct cftype files[] = {
 		.name = "stat",
 		.seq_show = cpuacct_stats_show,
 	},
+
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+	{
+		.name = "normalize_usage",
+		.read_u64 = cpuusage_normalize_read,
+		.write_u64 = cpuusage_normalize_write,
+	},
+#endif
 	{ }	/* terminate */
 };
 
-- 
2.17.1

